File: ruby/binary-search-tree/binary_search_tree.rb
class Bst
  attr_reader :data, :left, :right

  def initialize(n)
    @data = n
    @left = nil
    @right = nil
  end

  def insert(n)
    if n <= @data
      if @left
        @left.insert(n)
      else
        @left = Bst.new(n)
      end
    else
      if @right
        @right.insert(n)
      else
        @right = Bst.new(n)
      end
    end
  end

  def each(&block)
    if block_given?
      @left.each(&block) if @left
      yield @data
      @right.each(&block) if @right
    else
      enum_for(:each)
    end
  end
end

File: javascript/binary-search-tree/binary-search-tree.js
export class BinarySearchTree {
    constructor(n) {
        this.data = n;
        this.right = null;
        this.left = null;
    }

    insert(n) {
        if (n <= this.data) {
            if (this.left)
                this.left.insert(n);
            else
                this.left = new BinarySearchTree(n);
        } else {
            if (this.right)
                this.right.insert(n);
            else
                this.right = new BinarySearchTree(n);
        }
    }

    each(func) {
        if (this.left)
            this.left.each(x => func(x));

        func(this.data);

        if (this.right)
            this.right.each(x => func(x));
    }
}
